addons/loyalty/models/loyalty_program.py

```diff
@@ -0,0 +1,318 @@
+
+    _sql_constraints = [
+        ('check_max_usage', 'CHECK (limit_usage = False OR max_usage > 0)',
+            'Max usage must be strictly positive if a limit is used.'),
+    ]
+
+    @api.constrains('reward_ids')
+    def _constrains_reward_ids(self):
+        if any(not program.reward_ids for program in self):
+            raise ValidationError(_('A program must have at least one reward.'))
+
+    def _compute_total_order_count(self):
+        self.total_order_count = 0
+
+    @api.depends('company_id')
+    def _compute_currency_id(self):
+        for program in self:
+            program.currency_id = program.company_id.currency_id or program.currency_id
+
+    @api.depends('coupon_ids')
+    def _compute_coupon_count(self):
+        read_group_data = self.env['loyalty.card'].read_group([('program_id', 'in', self.ids)], ['program_id'], ['program_id'])
+        count_per_program = {r['program_id'][0]: r['program_id_count'] for r in read_group_data}
+        for program in self:
+            program.coupon_count = count_per_program.get(program.id, 0)
+
+    @api.depends('program_type', 'applies_on')
+    def _compute_is_nominative(self):
+        for program in self:
+            program.is_nominative = program.applies_on == 'both' or\
+                (program.program_type == 'ewallet' and program.applies_on == 'future')
+
+    @api.model
+    def _program_type_default_values(self):
+        # All values to change when program_type changes
+        # NOTE: any field used in `rule_ids`, `reward_ids` and `communication_plan_ids` MUST be present in the kanban view for it to work properly.
+        return {
+            'coupons': {
+                'applies_on': 'future',
+                'trigger': 'with_code',
+                'portal_visible': False,
+                'portal_point_name': 'Points',
+                # Coupons don't use rules by default
+                'rule_ids': [(5,)],
+                'reward_ids': [(5,), (0, 0, {
+                    'required_points': 1,
+                    'discount': 10,
+                })],
+                'communication_plan_ids': [(5,), (0, 0, {
+                    'trigger': 'create',
+                    'mail_template_id': (self.env.ref('loyalty.mail_template_loyalty_card', raise_if_not_found=False) or self.env['mail.template']).id,
+                })],
+            },
+            'promotion': {
+                'applies_on': 'current',
+                'trigger': 'auto',
+                'portal_visible': False,
+                'portal_point_name': 'Points',
+                'rule_ids': [(5,), (0, 0, {
+                    'reward_point_amount': 1,
+                    'reward_point_mode': 'unit',
+                })],
+                'reward_ids': [(5,), (0, 0, {
+                    'required_points': 1,
+                    'discount': 10,
+                    'clear_wallet': True
+                })],
+                'communication_plan_ids': [(5,)],
```

Comment:
Purely cosmetic / subjective / not mandatory

```suggestion
                'communication_plan_ids': [(5, 0, 0)],
```